projects/components/src/filtering/filter/filter.service.ts

```diff
@@ -0,0 +1,205 @@
+import { AbstractFilterBuilder } from './builder/types/abstract-filter-builder';
+import { Filter, IncompleteFilter } from './filter';
+import { FilterAttribute } from './filter-attribute';
+import { FilterOperator, fromUrlFilterOperator, toUrlFilterOperator } from './filter-operators';
+import { FilterParserLookupService } from './parser/filter-parser-lookup.service';
+import { SplitFilter } from './parser/parsed-filter';
+import { AbstractFilterParser } from './parser/types/abstract-filter-parser';
+
+@Injectable({
+  providedIn: 'root'
+})
+export class FilterService {
+  private static readonly FILTER_QUERY_PARAM: string = 'filter';
+
+  public constructor(
+    private readonly navigationService: NavigationService,
+    private readonly filterBuilderLookupService: FilterBuilderLookupService,
+    private readonly filterParserLookupService: FilterParserLookupService
+  ) {}
+
+  /*
+   * URL Filtering
+   */
+
+  public getUrlFiltersChanges$(attributes: FilterAttribute[]): Observable<Filter[]> {
+    return this.navigationService.navigation$.pipe(map(() => this.getUrlFilters(attributes)));
+  }
+
+  public getUrlFilters(attributes: FilterAttribute[]): Filter[] {
+    return this.navigationService
+      .getAllValuesForQueryParameter(FilterService.FILTER_QUERY_PARAM)
+      .map(filterString => this.parseUrlFilterString(attributes, filterString))
+      .filter((filter): filter is Filter => filter !== undefined);
+  }
+
+  public setUrlFilters(filters: Filter[]): void {
+    this.navigationService.addQueryParametersToUrl({
+      [FilterService.FILTER_QUERY_PARAM]: filters.length === 0 ? undefined : filters.map(f => f.urlString)
+    });
+  }
+
+  public applyUrlFilter(attributes: FilterAttribute[], filter: Filter): void {
+    const filters = this.getUrlFilters(attributes);
+
+    const foundIndex = filters.findIndex(f => filter.field === f.field);
+
+    if (foundIndex !== -1) {
+      filters[foundIndex] = filter;
+    } else {
+      filters.push(filter);
+    }
+
+    this.setUrlFilters([...filters]);
+  }
+
+  public removeUrlFilter(attributes: FilterAttribute[], filter: Filter): void {
+    this.setUrlFilters([...this.getUrlFilters(attributes).filter(f => filter.field !== f.field)]);
+  }
+
+  private parseUrlFilterString(attributes: FilterAttribute[], filterString: string): Filter | undefined {
+    return attributes
+      .filter(attribute => this.filterBuilderLookupService.isBuildableType(attribute.type))
+      .flatMap(attribute => {
+        const filterBuilder = this.filterBuilderLookupService.lookup(attribute.type);
+        const supportedUrlOperators = filterBuilder.supportedOperators().map(toUrlFilterOperator);
+
+        const splitUrlFilter = AbstractFilterParser.splitFilterStringByOperator(
+          supportedUrlOperators,
+          filterString,
+          false
+        );
+
+        if (splitUrlFilter === undefined) {
+          return undefined;
```

Comment:
does flatmap handle undefined return? would expect empty array

Reply:
It just creates an undefined element. We then filter those all out as the final step of this chain in order to grab what we found to be valid.